transport/src/main/java/io/netty/channel/local/LocalChannel.java

```diff
@@ -328,16 +330,37 @@ private void finishPeerRead(final LocalChannel peer) {
         if (peer.eventLoop() == eventLoop() && !peer.writeInProgress) {
             finishPeerRead0(peer);
         } else {
-            peer.eventLoop().execute(new OneTimeTask() {
-                @Override
-                public void run() {
-                    finishPeerRead0(peer);
-                }
-            });
+            runFinishPeerReadTask(peer);
+        }
+    }
+
+    private void runFinishPeerReadTask(final LocalChannel peer) {
+        // If the peer is writing, we must wait until after reads are completed for that peer before we can read. So
+        // we keep track of the task, and coordinate later that our read can't happen until the peer is done.
+        final Runnable finishPeerReadTask = new OneTimeTask() {
+            @Override
+            public void run() {
+                finishPeerRead0(peer);
+            }
+        };
+        if (peer.writeInProgress) {
+            peer.finishReadFuture = peer.eventLoop().submit(finishPeerReadTask);
+        } else {
+            peer.eventLoop().execute(finishPeerReadTask);
         }
     }
 
-    private static void finishPeerRead0(LocalChannel peer) {
+    private void finishPeerRead0(LocalChannel peer) {
+        Future<?> peerReadFuture = peer.finishReadFuture;
+        if (peerReadFuture != null) {
+            if (!peerReadFuture.isDone()) {
+                runFinishPeerReadTask(peer);
+                return;
+            } else {
+                // Lazy unset to make sure we don't prematurely unset it while scheduling a new task.
+                peer.finishReadFuture = null;
```

Comment:
would we need to check if the peer.finishReadFuture ==  peerReadFuture and only set to null in this case ? A.k.a an atomic operation.